x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java

```diff
@@ -134,16 +133,16 @@ private static List<String> extractEffectiveRoleNames(@Nullable AuthorizationInf
         if (authorizationInfo == null) {
             return null;
         }
-        final Role role = RBACEngine.maybeGetRBACEngineRole(authorizationInfo);
-        if (role == Role.EMPTY) {
-            return List.of();
-        } else {
-            final Map<String, Object> info = authorizationInfo.asMap();
-            if (false == info.containsKey(PRINCIPAL_ROLES_FIELD_NAME)) {
-                return null;
-            }
-            return Arrays.stream((String[]) info.get(PRINCIPAL_ROLES_FIELD_NAME)).sorted().toList();
+
+        final Map<String, Object> info = authorizationInfo.asMap();
+        final Object roleNames = info.get(PRINCIPAL_ROLES_FIELD_NAME);
+        // AuthorizationInfo from custom authorization engine may not have this field or have it as a different data type
+        if (false == roleNames instanceof String[]) {
+            assert false == authorizationInfo instanceof RBACEngine.RBACAuthorizationInfo
+                : "unexpected user.roles field [" + roleNames + "] for RBACAuthorizationInfo";
+            return null;
         }
+        return Arrays.stream((String[]) roleNames).sorted().toList();
```

Comment:
Refactor since we can now just rely on the value of `PRINCIPAL_ROLES_FIELD_NAME` instead of checking `Role.EMPTY` for special handling.